security/class-user-last-seen.php

```diff
@@ -0,0 +1,420 @@
+			} );
+			return;
+		}
+
+		$url = remove_query_arg( array(
+			'action',
+			'user_id',
+			'reset_last_seen_nonce',
+		) );
+
+		$url = add_query_arg( array(
+			'reset_last_seen_success' => 1,
+		), $url );
+
+		wp_safe_redirect( $url );
+		exit();
+	}
+
+	public function ignore_inactivity_check_for_user( $user_id, $until_timestamp = null ) {
+		if ( ! $until_timestamp ) {
+			$until_timestamp = strtotime( '+2 days' );
+		}
+
+		return update_user_meta( $user_id, self::LAST_SEEN_IGNORE_INACTIVITY_CHECK_UNTIL_META_KEY, $until_timestamp );
+	}
+
+	public function user_promoted( $user_id ) {
+		$user = get_userdata( $user_id );
+		if ( ! $user ) {
+			throw new \Exception( 'User not found' );
+		}
+
+		if ( ! $this->user_with_elevated_capabilities( $user ) ) {
+			return;
+		}
+
+		$this->ignore_inactivity_check_for_user( $user_id );
+	}
+
+	public function register_release_date() {
+		if ( ! wp_doing_ajax() && ! get_option( self::LAST_SEEN_RELEASE_DATE_TIMESTAMP_OPTION_KEY ) ) {
+			// Right after the first admin_init, set the release date timestamp
+			// to be used as a fallback for users that never logged in before.
+			add_option( self::LAST_SEEN_RELEASE_DATE_TIMESTAMP_OPTION_KEY, time(), '', 'no' );
+		}
+	}
+
+	public function is_considered_inactive( $user_id ) {
+		if ( ! $this->should_check_user_last_seen( $user_id ) ) {
+			return false;
+		}
+
+		$ignore_inactivity_check_until = get_user_meta( $user_id, self::LAST_SEEN_IGNORE_INACTIVITY_CHECK_UNTIL_META_KEY, true );
+		if ( $ignore_inactivity_check_until && $ignore_inactivity_check_until > time() ) {
+			return false;
+		}
+
+		$last_seen_timestamp = get_user_meta( $user_id, self::LAST_SEEN_META_KEY, true );
+		if ( $last_seen_timestamp ) {
+			return $last_seen_timestamp < $this->get_inactivity_timestamp();
+		}
+
+		$release_date_timestamp = get_option( self::LAST_SEEN_RELEASE_DATE_TIMESTAMP_OPTION_KEY );
+		if ( $release_date_timestamp ) {
+			return $release_date_timestamp < $this->get_inactivity_timestamp();
+		}
+
+		// Release date is not defined yet, so we can't consider the user inactive.
+		return false;
+	}
+
+	private function get_inactivity_timestamp() {
+		$days = constant( 'VIP_SECURITY_CONSIDER_USERS_INACTIVE_AFTER_DAYS' ) ? absint( constant( 'VIP_SECURITY_CONSIDER_USERS_INACTIVE_AFTER_DAYS' ) ) : 90;
+
+		return strtotime( sprintf( '-%d days', $days ) ) + self::LAST_SEEN_UPDATE_USER_META_CACHE_TTL;
+	}
+
+	private function is_block_action_enabled() {
+		return defined( 'VIP_SECURITY_INACTIVE_USERS_ACTION' ) && constant( 'VIP_SECURITY_INACTIVE_USERS_ACTION' ) === 'BLOCK';
+	}
+
+	private function should_check_user_last_seen( $user_id ) {
+		/**
+		 * Filters the users that should be skipped when checking/recording the last seen.
+		 *
+		 * @param array $skip_users The list of user IDs to skip.
+		 */
+		$skip_users = apply_filters( 'vip_security_last_seen_skip_users', array() );
```

Comment:
What do you think about moving this filter to `is_considered_inactive()`? That gives more flexibility for filtering the results of the check and might be more predictable.